### PR TITLE
Enhance short lived container autodiscovery

### PR DESCRIPTION
**Pull request by mfpierre**
### What does this PR do?

* Rework the docker config provider to avoid listing all containers too often
* Add delay in service & docker config removal for better short-lived container detection

### Motivation

What inspired you to submit this pull request?

### Additional Notes

e2e ✅

### CODE REVIEW

File: pkg/autodiscovery/providers/docker.go [hunk 1]
```diff
@@ -25,16 +26,22 @@ const (
...
+	syncCounter  int
 }
 
 // NewDockerConfigProvider returns a new ConfigProvider connected to docker.
 // Connectivity is not checked at this stage to allow for retries, Collect will do it.
 func NewDockerConfigProvider(config config.ConfigurationProviders) (ConfigProvider, error) {
-	return &DockerConfigProvider{}, nil
+	return &DockerConfigProvider{
+		// periodically resync if we're missing events
+		syncInterval: 30,
```
Comment (hkaj) [inline pkg/autodiscovery/providers/docker.go]: can we make it clearer (either in the comment or in the variable name) that the resync happens every 30 runs, not 30 sec?

File: pkg/autodiscovery/providers/docker.go [hunk 2]
```diff
@@ -53,12 +60,21 @@ func (d *DockerConfigProvider) Collect() ([]integration.Config, error) {
 		go d.listen()
 	}
 
-	containers, err := d.dockerUtil.AllContainerLabels()
-	if err != nil {
-		return []integration.Config{}, err
+	var containers map[string]map[string]string
+	// on the first run we collect all labels, then rely on events
```
Comment (ajacquemot) [inline pkg/autodiscovery/providers/docker.go]: I think think we should explain **why**

**Review by ajacquemot — COMMENTED** [review]
LGTM, thanks for working on this

**Comment by ajacquemot** [comment]
This should help for https://github.com/DataDog/datadog-agent/issues/3955

**Review by ajacquemot — APPROVED** [review]

**Review by hkaj — APPROVED** [review]

**Pull request merged**